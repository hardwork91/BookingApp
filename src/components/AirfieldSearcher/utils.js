/* this fuctions receives the current selected airfields and the data received from the search airfield API
 * and merges into a single array to be iterated by the functions that renders the select options.
 * Notice that spred operator is used with the two params, that is because a copy of them must be made,
 * so, the operatios "delete" and "splice", does not modify the original "selectedData" and "data" values.
 */
export const mergeData = ({ ...selectedData }, [...data]) => {
  data.map(({ apicode }) => {
    // if data element exists on selectedData, it must be deleted form selectedData in order to avoid being duplicated
    if (selectedData[apicode]) {
      delete selectedData[apicode];
    }
  });
  // then, the remains elements in selectedData are add at the top of the array
  Object.entries(selectedData).map(element => {
    const [apicode, rest] = element;
    data.splice(0, 0, { ...rest, apicode });
  });
  return data;
};
