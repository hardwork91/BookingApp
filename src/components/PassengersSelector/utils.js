import React from "react";
import { Typography, Popover, Icon } from "antd";

const { Text } = Typography;

/* displayed label of passengers selector control changes acording the selected categories.
 * this functions iterates over the current selected passengers and defines which is the propertly label to use.
 * gramatical sintaxis could be better, but we decide to keep it as simple as possible.
 */
export const generatePassengerSelectorLabel = passengers => {
  let categoriesWithAmount = 0;
  let totalAmount = 0;
  let singleCategory;
  for (const category in passengers) {
    const amount = passengers[category];
    if (amount > 0) {
      // check if there are more than one category selected
      categoriesWithAmount++;
      singleCategory = category;
    }
    totalAmount += amount;
  }

  // For correct gramatical sintaxis shuld be propertly to use formatMessage form React-intl.And usig formatters like "plural".
  // In this exaple we are not using it in order to keep the aplication simple due to internacionalization is not mandatory.
  let label =
    categoriesWithAmount > 1
      ? `${totalAmount} Travelers`
      : `${totalAmount} ${singleCategory}`;

  // if there is any category selected, the font color is marked in red. Alsoa a warning message is displayed over the componet to help color blind users.
  if (categoriesWithAmount === 0) {
    label = (
      <Popover
        content={
          <span>
            <Icon
              type="warning"
              style={{ color: "#f5222d", marginRight: "10px" }}
            />
            You must select at least one passenger
          </span>
        }
        trigger="click"
        defaultVisible
      >
        <Text type="danger">{`${totalAmount} Travelers`}</Text>
      </Popover>
    );
  }

  return {
    label,
    amount: totalAmount,
  };
};
