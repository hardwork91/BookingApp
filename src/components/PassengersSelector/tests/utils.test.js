import React from "react";
import renderer from "react-test-renderer";
import { generatePassengerSelectorLabel } from "../utils";

it("when generatePassengerSelectorLabel", () => {
  const passengers = { Child: 0, Adults: 1 };
  const tree = renderer
    .create(generatePassengerSelectorLabel(passengers).label)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("when generatePassengerSelectorLabel and have more than one category selected", () => {
  const passengers = { Child: 3, Adults: 1 };
  const label = generatePassengerSelectorLabel(passengers).label;
  const tree = renderer.create(label).toJSON();
  expect(tree).toMatchSnapshot();
});

// it("when generatePassengerSelectorLabel and does not have any pssenger selected", () => {
//   const passengers = { Child: 0, Adults: 0 };
//   const label = generatePassengerSelectorLabel(passengers).label;
//   const tree = renderer
//     .create(
//       <a className="ant-dropdown-link" onClick={e => e.preventDefault()}>
//         {label}
//       </a>
//     )
//     .toJSON();
//   expect(tree).toMatchSnapshot();
// });
