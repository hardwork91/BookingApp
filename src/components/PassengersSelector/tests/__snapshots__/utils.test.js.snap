// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when generatePassengerSelectorLabel 1`] = `"1 Adults"`;

exports[`when generatePassengerSelectorLabel and have more than one category selected 1`] = `"4 Travelers"`;
